Allow custom metadata for replay file

marcelotrevisani opened the pull request:
Allow custom metadata in replay files
    
Refactor replay metadata handling with dataclasses. Introduced `ReplayTestMetadata` dataclass for cleaner metadata management. Updated serialization and deserialization logic to use the new dataclass. Added new tests to validate metadata handling and functionality.

Closes #78

marcelotrevisani commented:
that is ready for review please
@nicoddemus @tadeu @prusse-martin

nicoddemus commented on README.rst:
```diff
@@ -69,6 +69,38 @@ execute the tests in the same order with::
 
+Additional metadata
+-------------------
+
```
Great docs!

```suggestion

*Version added: 1.6*

```

nicoddemus commented on src/pytest_replay/__init__.py:
```diff
@@ -39,6 +44,24 @@ def pytest_addoption(parser):
 
 
+@dataclasses.dataclass
+class ReplayTestMetadata:
```
This class is not only responsible for the metadata, but everything else related to the plugin, like start, outcome, etc. To avoid confusing it with the "metadata" attribute, how about naming it differently, say:

```suggestion
class ReplayTestInfo:
```

marcelotrevisani replied:
I agree, ``ReplayTestInfo`` is better, indeed

nicoddemus commented on src/pytest_replay/__init__.py:
```diff
@@ -39,6 +44,24 @@ def pytest_addoption(parser):
+        return {k: v for k, v in asdict(self).items() if v}
+
+
+class _ReplayTestMetadataDefaultDict(collections.defaultdict):
```
As per the other comment, for consistency:

```suggestion
class _ReplayTestInfoDefaultDict(collections.defaultdict):
```

nicoddemus commented on src/pytest_replay/__init__.py:
```diff
@@ -119,7 +142,10 @@ def pytest_collection_modifyitems(self, items, config):
                 # Ignore blank linkes and comments. (#70)
                 if stripped and not stripped.startswith(("#", "//")):
-                    nodeid = json.loads(stripped)["nodeid"]
+                    node_metadata = json.loads(stripped)
+                    nodeid = node_metadata["nodeid"]
+                    if "finish" in node_metadata:
+                        self.nodes[nodeid] = ReplayTestMetadata(**node_metadata)
```
```suggestion
                    node_info = json.loads(stripped)
                    nodeid = node_info["nodeid"]
                    if "finish" in node_info:
                        self.nodes[nodeid] = ReplayTestInfo(**node_info)
```

nicoddemus commented on tests/test_replay.py:
```diff
@@ -274,6 +274,68 @@ def test_filter_out_tests_not_in_file(testdir):
+    assert result.ret != 0
+
+    contents = [
+        json.loads(s)
+        for replay_file in (".pytest-replay-gw0.txt", ".pytest-replay-gw1.txt")
+        for s in (dir / replay_file).read_text().splitlines()
+    ]
+    contents.sort(key=lambda x: x["nodeid"])
+    assert contents[1]["metadata"]["seed"] < 100
+    assert (
+        len(
+            {
+                val["metadata"]["seed"]
+                for val in contents
+                if val.get("metadata", {}).get("seed")
+            }
+        )
+        > 1
+    )
+
+
+def test_replay_extra_metadata_load(pytester, tmp_path):
+    pytester.makepyfile(
+        """
+        import pytest
+
+        def test_load(replay_metadata):
+            assert replay_metadata.metadata == {"seed": 1234}
+        """
+    )
+    pytester.maketxtfile(
+        """{\"nodeid\": \"test_replay_extra_metadata_load.py::test_load\", \"start\": 1.0}
+    {\"nodeid\": \"test_replay_extra_metadata_load.py::test_load\", \"start\": 1.0, \"finish\": 2.0, \"outcome\": \"passed\", \"metadata\": {\"seed\": 1234}}
+    """
+    )
+    result = pytester.runpytest(
+        f"--replay={pytester.path / 'test_replay_extra_metadata_load.txt'}"
+    )
+    assert result.ret == 0
+    result.assert_outcomes(passed=1)
```
I think we can unify those two tests into one:

```suggestion
def test_metadata(pytester, tmp_path):
    pytester.makepyfile(
        """
        import pytest

        @pytest.fixture
        def seed(replay_metadata):
            assert replay_metadata.metadata == {}
            replay_metadata.metadata["seed"] = seed = 1234
            return seed
        
        def test_foo(seed):
            assert seed == 1234
        """
    )
    dir = tmp_path / "replay"
    result = pytester.runpytest(f"--replay-record-dir={dir}")
    assert result.ret == 0

    # Rewrite the fixture to always returns the metadata, as written previously.
    pytester.makepyfile(
        """
        import pytest

        @pytest.fixture
        def seed(replay_metadata):
            return replay_metadata.metadata["seed"]            
        
        def test_foo(seed):
            assert seed == 1234
        """
    )
    result = pytester.runpytest(f"--replay={dir / '.pytest-replay.txt'}")
    assert result.ret == 0
```

This is simpler, simulates the actual workflow, and avoids having to write the replay data manually. 👍

nicoddemus reviewed (changes requested):
Great work!

Left some comments, please take a look.

prusse-martin commented on src/pytest_replay/__init__.py:
```diff
@@ -119,7 +142,10 @@ def pytest_collection_modifyitems(self, items, config):
                 # Ignore blank linkes and comments. (#70)
                 if stripped and not stripped.startswith(("#", "//")):
-                    nodeid = json.loads(stripped)["nodeid"]
+                    node_metadata = json.loads(stripped)
+                    nodeid = node_metadata["nodeid"]
+                    if "finish" in node_metadata:
+                        self.nodes[nodeid] = ReplayTestMetadata(**node_metadata)
```
I believe that if somehow the test crashes before the finish you metadata would no be save, but it could be relevant to the crash reproduction.

You possible want to emit to the replay file "update metadata" instructions so they could be loaded back to reproduce the problem.
Something in the lines of:
```json
{"nodeid": "test_1.py::test_bar", "start": 3.0}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "metadata": {"a": "asd"}}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "metadata": {"a": "asd", "b": 7}}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "finish": 4.0, "outcome": "passed"}
```
There we see the metadata evolving.

And probably could help pinpoint the crash source if the output is just:
```json
{"nodeid": "test_1.py::test_bar", "start": 3.0}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "metadata": {"a": "asd"}}
```
(a crash cause some lines to not be present)

marcelotrevisani replied:
Thanks for the feedback! You're absolutely right, and this is something I already noted in the issue under "Limitations". In my specific scenario, it's quite rare for tests to crash, they usually just fail, so the current approach works well enough for my use case. However, I do see the value in capturing incremental metadata.
To fully address this broader scenario, we'd likely need to track all setups and all teardowns associated with a test, ensuring the metadata reflects each setup and teardown, since a crash could happen at any point.

For now, I’d prefer to move forward with the current implementation, but I completely agree that your suggestion would be the ideal enhancement. It’s definitely something worth considering for future improvements! Maybe, could you create an issue about it so we can tackle that in the future or when someone sees values in it please?

marcelotrevisani commented:
> Great work!
> 
> Left some comments, please take a look.

Thank you for your feedback!

I have addressed all your comments in the review, could you please take a second look? Thanks!

nicoddemus reviewed (approved):
Nice, thanks!

After @prusse-martin's approval we can make a new release.
